src/entities/database.js

```diff
@@ -268,26 +305,170 @@ class DB implements Database {
+        validTxs.push(tx)
+      } else {
+        this.#logger.info(`tx ${tx} inputs already spent`)
+        invalidTxs.push(tx)
+      }
+    }
+    return [validTxs, invalidTxs]
+  }
+
+  async groupPendingTxsForSnapshot(txHashes: Array<string>): Promise<[Array<string>, Array<string>]> {
+    const pendingSet = await this.queryPendingSet()
+    const txsInPendingState = pendingSet.filter(hash => !txHashes.includes(hash))
+    const [pendingTxs, invalidTxs] = await this.groupPendingTxs(txsInPendingState)
+    return [pendingTxs, invalidTxs]
+  }
+
+
+  async storeNewPendingSnapshot(block: Block, snapshot: Array<string>) {
+    if (_.isEmpty(snapshot)) {
+      this.#logger.debug('storeNewPendingSnapshot: No pending txs added to snapshot..')
+      return
+    }
+    const dbFields = snapshot.map(txHash => ({
+      tx_hash: txHash,
+      block_hash: block.hash,
+      block_height: block.height,
+      status: TX_STATUS.TX_PENDING_STATUS,
+    }))
+    const sql = Q.sql.insert().into(SNAPSHOTS_TABLE)
+      .setFieldsRows(dbFields).toString()
+    this.#logger.debug('storeNewPendingSnapshot: ', snapshot, sql)
+    await this.getConn().query(sql)
+  }
+
+  async queryFailedSet() {
+    const sql = Q.sql.select().from('txs')
+      .field('hash')
+      .where('tx_state = ?', TX_STATUS.TX_FAILED_STATUS)
+      .where('NOT EXISTS ?', Q.sql.select().from(SNAPSHOTS_TABLE)
+        .where('status = ?', TX_STATUS.TX_FAILED_STATUS)
+        .where('tx_hash = hash'))
+      .toString()
+    const dbRes = await this.getConn().query(sql)
+    this.#logger.debug('queryFailedSet:', sql, dbRes)
+    return _.map(dbRes.rows, 'hash')
+  }
+
+  async storeNewFailedSnapshot(block: Block, invalidTxs: Array<string>) {
+    const failedSet = [
+      ...(await this.queryFailedSet()),
+      ...invalidTxs,
+    ]
+    if (_.isEmpty(failedSet)) {
+      this.#logger.debug('storeNewFailedSnapshot: No failed txs added to snapshot..')
+      return
+    }
+    const dbFields = failedSet.map(txHash => ({
+      tx_hash: txHash,
+      block_hash: block.hash,
+      block_height: block.height,
+      status: TX_STATUS.TX_FAILED_STATUS,
+    }))
+    const sql = Q.sql.insert().into(SNAPSHOTS_TABLE)
+      .setFieldsRows(dbFields).toString()
+    this.#logger.debug('storeNewFailedSnapshot: ', sql)
+    await this.getConn().query(sql)
+  }
+
+  async updateTxsStatus(txs: Array<string>, status: string) {
+    const sql = Q.sql.update().table('txs')
+      .set('tx_state', status)
+      .where('hash IN ?', txs)
+      .toString()
+    return this.getConn().query(sql)
+  }
+
+  async storeNewSnapshot(block: Block) {
+    const txHashes = _.map(block.txs, 'id')
+    const [pendingTxs, invalidTxs] = await this.groupPendingTxsForSnapshot(txHashes)
+    if (!_.isEmpty(invalidTxs)) {
+      await this.updateTxsStatus(invalidTxs, TX_STATUS.TX_FAILED_STATUS)
+    }
+    await this.storeNewPendingSnapshot(block, pendingTxs)
+    await this.storeNewFailedSnapshot(block, invalidTxs)
```

Comment:
I think there's no need to pass `invalidTxs` here into the `storeNewFailedSnapshot` because that function combines it with all new failed records it finds in the `txs` table, and because here we first call `updateTxsStatus` - it means that the same set of tx hashes will be selected from DB, so we are rather duplicating the lists here and are at risk of trying to write the same tx_hashes into the transient table at once.

Reply:
> here we first call updateTxsStatus - it means that the same set of tx hashes will be selected from DB
this update will not be available for `storeNewPendingSnapshot`, because these two queries are executed in the same db transaction.